src/statement.js

```diff
@@ -47,6 +47,19 @@ pp.isLet = function() {
   return false
 }
 
+// check 'async [no LineTerminator here] function'
+// - 'async /*foo*/ function' is OK.
+// - 'async /*\n*/ function' is invalid.
+pp.isAsyncFunction = function() {
+  if (this.type !== tt.name || this.options.ecmaVersion < 8 || this.value != "async")
+    return false
+
+  skipWhiteSpace.lastIndex = this.pos
+  let skip = skipWhiteSpace.exec(this.input)
+  let next = this.pos + skip[0].length
+  return !lineBreak.test(this.input.slice(this.pos, next)) && this.input.slice(next, next + 8) === "function"
```

Comment:
This should probably check for a non-identifier character after `function`, to make sure that the next token is actually the keyword `function`, not, say `functioning`.